File: books/cs/general-purpose-algorithms/introduction-to-algorithms/3rd-edition/chapter4/C/find_maximum_subarray.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <math.h>
#include <float.h>

#define DATASETSIZE 10

struct rtnData{
	int max_left;
	int max_right;
	double sum;
};

struct rtnData find_maximum_subarray(double* srcdata, int low, int high);
void printArray(double* srcdata){
	int idx;

    for( idx=0; idx<DATASETSIZE; idx++ ){
        if(idx%10 == 0) printf("\n\n");
        printf(" %f, ",srcdata[idx]);
    }
    printf("\n");
}

int main(void){
    double srcdata[DATASETSIZE];
	struct rtnData rtn;
    int idx, jdx, sign;

    srand(time(NULL));
    for( idx=0; idx<DATASETSIZE; idx++ ){
        srcdata[idx] = rand()*1.0/RAND_MAX;
		sign = ( srcdata[idx] > 0.5 )?-1:1;
		srcdata[idx] *= sign;
    }
	printArray(srcdata);

	rtn = find_maximum_subarray(srcdata, 0, DATASETSIZE-1);
	printf("(left,right,sum) = (%d,%d,%f)\n",rtn.max_left,rtn.max_right,rtn.sum);
	printf("DBL_MAX = %f\n",-1*DBL_MAX);
	return 0;
}

struct rtnData find_max_crossing_subarray(double* srcdata, int low, int mid, int high){
	struct rtnData rtn;
	double sum,left_sum,right_sum;
	int idx;
	
	sum = 0;
	left_sum = -1*DBL_MAX;
	for(idx=mid; idx>=low;idx--){
		sum += srcdata[idx];
		if( sum > left_sum ){
			left_sum = sum;
			rtn.max_left = idx;
		}
	}
	
	sum = 0;
	right_sum = -1*DBL_MAX;
	for(idx=mid+1;idx<=high;idx++){
		sum += srcdata[idx];
		if( sum>right_sum ){
			right_sum = sum;
			rtn.max_right = idx;
		}
	}
	rtn.sum = left_sum + right_sum;
	return rtn;
}

struct rtnData find_maximum_subarray(double* srcdata, int low, int high){
	struct rtnData rtn,rtn_left,rtn_right,rtn_crossing;
	int mid=0;

	if( high == low ){
		rtn.max_left = low;
		rtn.max_right = high;
		rtn.sum = srcdata[low];
	}else{
		mid = (int)floor((low+high)/2);
		rtn_left = find_maximum_subarray(srcdata,low,mid);
		rtn_right = find_maximum_subarray(srcdata,mid+1,high);
		rtn_crossing = find_max_crossing_subarray(srcdata,low,mid,high);

		if( (rtn_left.sum >= rtn_right.sum) && (rtn_left.sum >= rtn_crossing.sum) ){
			rtn = rtn_left;
		}else if ( (rtn_right.sum >= rtn_left.sum) && (rtn_right.sum >= rtn_crossing.sum) ){
			rtn = rtn_right;
		}else{
			rtn = rtn_crossing;
		}
	}
	
	return rtn;	
}

File: books/cs/general-purpose-algorithms/introduction-to-algorithms/3rd-edition/chapter4/Scala/find_maximum_subarray.scala
import scala.util.Random
import scala.collection._

object HeapSort{
    def main(args: Array[String]):Unit = {
        val dataSetSize = 10
        var srcdata = util.Random.shuffle(100 to 999).take(dataSetSize).toArray;
        var aaa = new Array[Int](dataSetSize)
        var bbb = new Array[Int](aaa.size)
        val digit = 3
        var idx:Int = 0
        var jdx:Int = 0
		var rtn: (Int,Int,Int) = (0,0,0)

        println(s"the source array, ${srcdata.toList}")
        //heap_sort(srcdata,dataSetSize)
		rtn = find_max_subarray( srcdata, 1, dataSetSize )
        println(s"the sorted data, ${srcdata.toList}")
    }

    def swap(srcdata: Array[Int], p: Int, r: Int):Unit = {
        var temp = srcdata(p)
        srcdata(p) = srcdata(r)
        srcdata(r) = temp
    }

    def left( i: Int ): Int = {
        return(2*i)
    }
    def right( i: Int ): Int = {
        return(2*i+1)
    }


    def max_heapify(srcdata: Array[Int], len: Int, i: Int):Unit = {
        val l = left(i)
        val r = right(i)
        var largest = 0

        if( l<len && srcdata(l)>srcdata(i) ){
            largest = l
        }else largest = i
        if( r<len && srcdata(r)>srcdata(largest) ){
            largest = r
        }

        if( largest != i ){
            swap(srcdata,largest,i)
            max_heapify(srcdata,len,largest)
        }

    }

    def build_max_heap(srcdata: Array[Int], len: Int):Unit = {
        var idx: Int = 0

        for( idx <- len/2 to 0 by -1 ){
            max_heapify(srcdata,len,idx)
        }
    }

    def heap_sort(srcdata: Array[Int], len: Int):Unit = {
        var idx: Int = 0
        var datasize = len

        build_max_heap(srcdata,len)
        for( idx <- len-1 to 1 by -1 ){
            swap(srcdata,0,idx)
            datasize = datasize -1
            max_heapify(srcdata,datasize,0)
        }
    }

	def find_max_crossing_subarray( srcdata: Array[Int], low: Int, mid: Int, high: Int ):(Int, Int, Int) = {
		var left_sum: Int = -1e9.toInt
		var right_sum: Int = -1e9.toInt
		var sum: Int = 0
		var max_left: Int = mid
		var max_right: Int = mid
		var idx: Int = 0
		for( idx <- mid to low by -1 ){
			sum = sum + srcdata(idx)
			if( sum > left_sum ){
				left_sum = sum
				max_left = idx
			}
		}
		
		sum = 0
		for( idx <- mid+1 to high ){
			sum = sum + srcdata(idx)
			if( sum > right_sum ){
				right_sum = sum
				max_right = idx
			}
		}
		
		return( (max_left,max_right,left_sum+right_sum) )
	}

	def find_max_subarray( srcdata: Array[Int], low: Int, high: Int ):(Int, Int, Int) = {
		var mid: Int = low
		var rtn: (Int, Int, Int) = (0,0,0)
		var rtn_left: (Int, Int, Int) = (0,0,0)
		var rtn_right: (Int, Int, Int) = (0,0,0)
		var rtn_crossing: (Int, Int, Int) = (0,0,0)

		if( low == high ){
			rtn = (low, high, srcdata(low))
		}else{
			mid = (low+high)/2	
			rtn_left = find_max_subarray(srcdata,low,mid)
			rtn_right = find_max_subarray(srcdata,mid+1,high)
			rtn_crossing = find_max_crossing_subarray(srcdata,low,mid,high)
			
			if( (rtn_left._3>rtn_right._3) && (rtn_left._3>rtn_crossing._3) ){
				rtn = rtn_left
			}else if( (rtn_right._3>rtn_left._3) && (rtn_right._3>rtn_crossing._3) ){
				rtn = rtn_right
			}else{
				rtn = rtn_crossing
			}
		}
		return(rtn)
	}
}
